src/Giraffe/Routing.fs

```diff
@@ -10,37 +10,89 @@ open FSharp.Control.Tasks.V2.ContextInsensitive
 open Giraffe.FormatExpressions
 
 // ---------------------------
-// Private sub route helper functions
+// Sub Routing Feature
 // ---------------------------
 
-[<RequireQualifiedAccess>]
-module SubRouting =
+[<Literal>]
+let private RouteKey = "giraffe_route"
+
+type ISubRoutingFeature =
+    abstract member GetResolvedPath   : unit   -> string option
+    abstract member SetResolvedPath   : string -> unit
+    abstract member ClearResolvedPath : unit   -> unit
+
+type SubRoutingFeature() =
+    let mutable resolvedPath = None
 
-    [<Literal>]
-    let private RouteKey = "giraffe_route"
+    interface ISubRoutingFeature with
+        member __.GetResolvedPath()             = resolvedPath
+        member __.SetResolvedPath (p : string)  = resolvedPath <- Some p
+        member __.ClearResolvedPath()           = resolvedPath <- None
 
-    let getSavedPartialPath (ctx : HttpContext) =
-        if ctx.Items.ContainsKey RouteKey
-        then ctx.Items.Item RouteKey |> string |> strOption
+type HttpContext with
+
+    [<Obsolete("This method has been deprecated. Please use 'GetResolvedPath' instead which is a more memory efficient implementation based on the HttpContext.Features API.")>]
+    member this.GetResolvedPathFromItems() =
+        if this.Items.ContainsKey RouteKey
+        then this.Items.Item RouteKey |> string |> strOption
         else None
 
-    let getNextPartOfPath (ctx : HttpContext) =
-        match getSavedPartialPath ctx with
-        | Some p when ctx.Request.Path.Value.Contains p -> ctx.Request.Path.Value.[p.Length..]
-        | _   -> ctx.Request.Path.Value
+    /// **Description**
+    ///
+    /// Returns a partially (or fully) resolved request path during routing decisions when working with the SubRoutingFeature.
+    ///
+    member this.GetResolvedPath() =
+        let subRoutingFeature = this.Features.Get<ISubRoutingFeature>()
+        match subRoutingFeature.GetResolvedPath() with
+        | Some p -> Some p
+        | None   -> this.GetResolvedPathFromItems()
+
+    /// **Description**
+    ///
+    /// Returns the next part of the request path, which hasn't been resolved by a routing handler yet.
+    ///
+    member this.GetNextPartOfPath() =
+        let requestPath = this.Request.Path.Value
+        match this.GetResolvedPath() with
+        // Should this be a StartsWith?
+        | Some p when requestPath.Contains p -> requestPath.[p.Length..]
+        | _ -> requestPath
+
+    /// **Description**
+    ///
+    /// Memorises a partially resolved path between different routing handlers by making use of the SubRoutingFeature.
+    ///
+    member this.SetResolvedPath (path : string) =
+        let subRoutingFeature = this.Features.Get<ISubRoutingFeature>()
+        subRoutingFeature.SetResolvedPath path
+        // Keep this as a fallback for Saturn, etc?
```

Comment:
Not sure if this is necessary for Saturn, feels counterintuitive to still populate the Items collection when using a feature now?

Reply:
I'll do a check, we may only interface with the routing apis, if not I'll submit a pr. If we do so because of an api gap in Giraffe we'll need a compatibility release so we can switch :)

Reply:
Here's the code that knows about the items key:
https://github.com/SaturnFramework/Saturn/blob/d4506da6721a592b4601a27f40eae870e7e59e15/src/Saturn/Common.fs#L35-L131

Reply:
Could be over engineering here, but we could adopt a similar approach that MVC did, and have a compatibility setting that is defaulted to let's say `3.6`, and consumers opt-in to setting it to `4.0` when they add Giraffe in the startup, then here we'd conditionally do the compatibility thing based on that?

Reply:
I've added a new `GiraffeCompatibilityMode` type which can be set when registerting the GiraffeMiddleware. I've also started to change existing sub route tests to run in both modes.  I think this is an elegant solution and it's actually a very small change. The most part is just slightly re-factoring existing tests, but that makes testing for this change easier too, so overall a good thing!